src/python/pants/backend/python/tasks/build_local_python_distributions.py

```diff
@@ -44,89 +45,103 @@ def product_types(cls):
     return True
 
-  @staticmethod
-  def filter_target(tgt):
-    return type(tgt) is PythonDistribution
+  def _get_setup_requires_to_resolve(self, dist_target):
+    if not dist_target.setup_requires:
+      return None
 
-  def _ensure_setup_requires_site_dir(self, dist_targets, interpreter, site_dir):
     reqs_to_resolve = set()
 
-    for tgt in dist_targets:
-      for setup_req_lib_addr in tgt.setup_requires:
-        for req_lib in self.context.build_graph.resolve(setup_req_lib_addr):
-          for req in req_lib.requirements:
-            reqs_to_resolve.add(req)
+    for setup_req_lib_addr in dist_target.setup_requires:
+      for req_lib in self.context.build_graph.resolve(setup_req_lib_addr):
+        for req in req_lib.requirements:
+          reqs_to_resolve.add(req)
 
-    if not reqs_to_resolve:
+    if len(reqs_to_resolve) == 0:
       return None
-    self.context.log.debug('python_dist target(s) with setup_requires detected. '
-                           'Installing setup requirements: {}\n\n'
-                           .format([req.key for req in reqs_to_resolve]))
 
-    setup_requires_dists = _resolve_multi(interpreter, reqs_to_resolve, ['current'], None)
+    return reqs_to_resolve
 
-    overrides = {
-      'purelib': site_dir,
-      'headers': os.path.join(site_dir, 'headers'),
-      'scripts': os.path.join(site_dir, 'bin'),
-      'platlib': site_dir,
-      'data': site_dir
-    }
+  # NB: these are all the immediate subdirectories of the target's results directory.
+  # This contains any modules from a setup_requires().
+  setup_requires_site_subdir = 'setup_requires_site'
+  # This will contain the sources used to build the python_dist().
+  dist_subdir = 'python_dist_subdir'
 
-    # The `python_dist` target builds for the current platform only.
-    for obj in setup_requires_dists['current']:
-      wf = WheelFile(obj.location)
-      wf.install(overrides=overrides, force=True)
+  @classmethod
+  def _get_output_dir(cls, results_dir):
+    return os.path.join(results_dir, cls.dist_subdir)
 
-    return site_dir
+  @classmethod
+  def _get_dist_dir(cls, results_dir):
+    return os.path.join(cls._get_output_dir(results_dir), SetupPyRunner.DIST_DIR)
 
   def execute(self):
-    dist_targets = self.context.targets(self.filter_target)
+    dist_targets = self.context.targets(is_local_python_dist)
 
     if dist_targets:
-      with self.invalidated(dist_targets,
-                            fingerprint_strategy=DefaultFingerprintStrategy(),
-                            invalidate_dependents=True) as invalidation_check:
-        interpreter = self.context.products.get_data(PythonInterpreter)
+      interpreter = self.context.products.get_data(PythonInterpreter)
+      shared_libs_product = self.context.products.get(SharedLibrary)
 
+      with self.invalidated(dist_targets, invalidate_dependents=True) as invalidation_check:
         for vt in invalidation_check.invalid_vts:
-          if vt.target.dependencies:
```

Comment:
It is worth mentioning that allowing dependencies here breaks the original vision for python_dist as a Pants-specific apparatus for depending on a properly-packaged and readily-consumable python distributions (that exists in the python ecosystem with nothing Pants-specific at all). Clearly it is crucial to this PR that we allow dependencies and I don't think it it is a problem, but I wanted to mention that we may be stepping on/adding additional conceptual loading onto what a "python distribution" is in Pants context. It may make sense to rename the target altogether or add some explicit docs/visibility to the fact that this target now offers more native integrations into Pants-land (via dependencies) that the original `python_dist` would not allow.

Reply:
I totally see this! I guess my thought was that: this feature doesn't seem to stop us from supporting a "properly-packaged and readily-consumable python distribution" (and that was part of the rationale for why it works like this). Because dependencies aren't allowed in a "normal" `python_dist()`, and because the only intrusion we have into the `setup.py` is adding `data_files=[('', ['libasdf-c.so', 'libasdf-cpp.so'])],` which seems like verbatim what you would add if you were doing externally-compiled artifacts without using Pants. The fact that it ends up setting the platform might be a little magic (and needs documentation), but again it also seems like what you would do if your goal was to externally compile some native artifacts and insert them into your `python_dist()`. This technique works *seamlessly* with native code in the `python_dist()` itself, because we're just adding a resource. Could you explain what concepts we're stepping on here? It's not clear to me.

If you mean "make it more clear the tradeoffs of this technique vs compiling native code in the `python_dist()`", I'm 100% with you, there are distinct use cases for both (using the Python C/C++ API vs accessing native code with `ctypes`), and that deserves a clear set of documentation. If there are other inconsistencies that should be clearly noted in a page on pantsbuild.org, let me know, I'll make a followup issue for that now.